static-build/posts/2021/09/cors/index.md

```diff
@@ -0,0 +1,348 @@
+
+If someone says they understand the security implications of a URL without UI intervention, be sure to check they can recite all 9000+ entries of the public suffix list from memory.
+
+# Opening things up again
+
+Ok, so we've got these APIs like `<img>` that can access resources from other origins, but visibility into the response is limited (but not limited enough in hindsight), and we've got these more powerful APIs like cross-frame scripting and `XMLHttpRequest` which only work same-origin.
+
+How could we allow those more powerful APIs to work _across_ origins?
+
+## Remove credentials?
+
+Let's say we provide an opt-in so the request is sent without credentials. The response will be the 'logged-out' view, so it won't contain any private data, and can be revealed without concern, right?
+
+Unfortunately there's a lot of HTTP endpoints out there that 'secure' themselves using things other than browser credentials.
+
+A lot of company intranets assume they're 'private' because they're only accessible from a particular network. Some routers and IoT devices assume they're only accessible by well-meaning folks because they're restricted to your home network (remember, the 's' in 'IoT' stands for security). Some websites offer different content depending on the IP address they're accessed from.
+
+So, if you visit my website from your home, I could start making requests to common hostnames and IP addresses, looking for unsecure IoT devices, looking for routers using default passwords, and generally make your life very miserable, all without needing browser credentials.
+
+Removing credentials is part of the solution, but it isn't enough on its own. There's no way to know that a resource contains private data without browser credentials, so we need some way for the resource to declare "hey, it's fine, let the other site read my content".
+
+## Separate resource opt-in
+
+The origin could have some special resource that details its permissions regarding cross-origin access. That's the [security model Flash went with](https://www.adobe.com/devnet-docs/acrobatetk/tools/AppSec/xdomain.html). Flash looked for a `/corssdomain.xml` in the root of the site that looked like this:
+
+```xml
+<?xml version="1.0"?>
+<!DOCTYPE cross-domain-policy SYSTEM "https://www.adobe.com/xml/dtds/cross-domain-policy.dtd">
+<cross-domain-policy>
+  <site-control permitted-cross-domain-policies="master-only" />
+  <allow-access-from domain="*.example.com" />
+  <allow-access-from domain="www.example.com" />
+  <allow-http-request-headers-from domain="*.adobe.com" headers="SOAPAction" />
+</cross-domain-policy>
+```
+
+Unfortunately this is a bit all-or-nothing. You can imagine a similar format that lets you allow-list particular resources, but the resource would start to get quite large. It also means this potentially large resource would need to be downloaded before downloading the actual thing you want.
```

Comment:
It also gets out-of-date with practice on the server quickly as it's maintained in a rather separate place.